compiler-rt/lib/scudo/standalone/combined.h

```diff
@@ -1334,17 +1337,25 @@ class Allocator {
 
     Chunk::storeHeader(Cookie, Ptr, Header);
 
-    if (BypassQuarantine) {
-      void *BlockBegin;
-      if (LIKELY(!useMemoryTagging<AllocatorConfig>(Options))) {
-        // Must do this after storeHeader because loadHeader uses a tagged ptr.
-        if (allocatorSupportsMemoryTagging<AllocatorConfig>())
-          Ptr = untagPointer(Ptr);
-        BlockBegin = getBlockBegin(Ptr, Header);
-      } else {
-        BlockBegin = retagBlock(Options, TaggedPtr, Ptr, Header, Size, true);
+    void *BlockBegin;
+    if (LIKELY(!useMemoryTagging<AllocatorConfig>(Options))) {
+      // Must do this after storeHeader because loadHeader uses a tagged ptr.
+      if (allocatorSupportsMemoryTagging<AllocatorConfig>())
+        Ptr = untagPointer(Ptr);
+      BlockBegin = getBlockBegin(Ptr, Header);
+    } else {
+      BlockBegin = retagBlock(Options, TaggedPtr, Ptr, Header, Size, true);
+    }
+
+    if (AllocatorConfig::getEnableZeroOnDealloc()) {
+      uptr length = reinterpret_cast<uptr>(Ptr) + Size -
+                    reinterpret_cast<uptr>(BlockBegin);
+      if (length <= ZeroOnDeallocMaxSize) {
+        memset(BlockBegin, 0, length);
       }
+    }
```

Comment:
`EnableZeroOnDealloc` is supposed to be only available on no-memory-tagging and no-quarantine (These two flags require the header) so we should put this in the if (BypassQuarantine) { ... }

Let's also add a DCHECK(!useMemoryTagging<...>(...));

Reply:
Good point. Done.